Remove hardcoded hotkey references from production, support powers, and viewport.

pchote opened the pull request:
Part 3 of my campaign to unhardcode and move gameplay-specific hotkeys to mod code.
This follows on from #13706 and #13711.

This converts all the remaining hotkey consumers except for `WorldCommandWidget` and the hardcoded "dev" hotkeys.  Those two require more drastic changes that will come in their own PRs.

pchote committed:
Move hardcoded production tab keys into yaml.

pchote committed:
Allow widgets to expose custom hotkey names to the linter.

pchote committed:
Move hardcoded production keys into yaml.

pchote committed:
Move hardcoded support power keys into yaml.

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/SupportPowerTooltipLogic.cs:
```diff
@@ -57,12 +57,23 @@ public SupportPowerTooltipLogic(Widget widget, TooltipContainerWidget tooltipCon
+					}
+
+					hotkeyLabel.GetText = () => hotkeyText;
+					hotkeyLabel.Bounds.X = nameFont.Measure(name).X + 2 * nameLabel.Bounds.X;
```
Lines hotkeyText, GetText, Bounds can be inside IsValid() block.

rob-v replied:
var nameWidth = font.Measure(name).X; could be used and reused below like in ProductionTooltipLogic.

rob-v commented on OpenRA.Mods.Common/Widgets/ProductionTabsWidget.cs:
```diff
@@ -68,7 +68,11 @@ public class ProductionTabsWidget : Widget
-		public Dictionary<string, ProductionTabGroup> Groups;
+
+		public readonly NamedHotkey NextProductionTabKey = new NamedHotkey();
+		public readonly NamedHotkey PreviousProductionTabKey = new NamedHotkey();
```
It could be Previous, Next like below and in yaml.

rob-v commented on OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs:
```diff
@@ -71,6 +87,20 @@ public void Run(Action<string> emitError, Action<string> emitWarning, ModData mo
+
+					Hotkey unused;
+					foreach (var name in keyNames)
+						if (!namedKeys.Contains(name) && !Hotkey.TryParse(name, out unused))
```
This works fine with HotkeyCount=0, but with missing HotkeyPrefix it passes lint but produces strange/wrong hotkeys. Ideally HotkeyPrefix should be mandatory for count>0.

rob-v reviewed (commented):
mostly optional nits.

pchote committed:
Move hardcoded viewport keys into yaml.

pchote committed:
Clean up ProductionTooltipLogic:

- Avoid creating unnecessary bindings
- Fix time display not updating on low power
- Fix power usage color not updating on power changes
- Avoid duplicated text size calculations

pchote committed:
Clean up SupportPowerTooltipLogic:

- Avoid creating unnecessary bindings
- Avoid duplicated text size calculations
- Relayout panel when (and only when) needed
- Color timer red when paused

pchote commented:
Updated as requested, and added a few more cleanups on top of your suggestions to the support/production tooltip logic.

rob-v commented on OpenRA.Mods.Common/Widgets/SupportPowersWidget.cs:
```diff
@@ -47,12 +52,35 @@ public class SupportPowersWidget : Widget
+				return new string[0];
+
+			if (string.IsNullOrEmpty(prefix))
+				throw new YamlException("ProductionPalette must define HotkeyPrefix if HotkeyCount > 0.");
```
SupportPowers/SupportPalette instead of ProductionPalette here.

rob-v commented on OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:
```diff
@@ -71,32 +82,53 @@ public class ViewportControllerWidget : Widget
 		{
-			SaveBookmark(index, worldRenderer.Viewport.CenterPosition);
-		}
+			var savePrefix = string.Empty;
```
You used `prefix = ""` in SupportPowersWidget, ProductionPaletteWidget.

rob-v commented on OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:
```diff
@@ -71,32 +82,53 @@ public class ViewportControllerWidget : Widget
+				throw new YamlException("Viewportcontroller must define BookmarkSaveKeyPrefix if BookmarkKeyCount > 0.");
+
+			if (string.IsNullOrEmpty(restorePrefix))
+				throw new YamlException("Viewportcontroller must define BookmarkSaveKeyPrefix if BookmarkKeyCount > 0.");
```
BookmarkRestoreKeyPrefix here.

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/SupportPowerTooltipLogic.cs:
```diff
@@ -18,64 +19,74 @@ namespace OpenRA.Mods.Common.Widgets.Logic
-				var timeWidth = timeFont.Measure(time).X;
-				var topWidth = nameFont.Measure(name).X + hotkeyWidth + timeWidth + timeOffset;
-				var descSize = descFont.Measure(desc);
+					hotkeyWidth = nameSize.X + 2 * nameLabel.Bounds.X;
```
nameFont.Measure(hotkeyText).X should be kept here?

rob-v reviewed (commented):
small issues

pchote commented:
D'oh.  Fixed.

rob-v reviewed (approved):
lgtm 👍 

throw in LinterHotkeyNames is less convinient - emitError could be passed to use instead or added try/catch around LinterHotkeyNames call - to display only mesage and continue checking instead of stacktrace and 'crash'/stop on first occurrence.

pchote commented:
Good idea, will do.

pchote commented:
Done.  This also let me print the file and line numbers:
```
Testing Tiberian Sun mod MiniYAML...
Testing mod: Tiberian Sun
OpenRA.Utility(1,1): Error: ts|chrome/ingame-player.yaml:474 must define HotkeyPrefix if HotkeyCount > 0.
OpenRA.Utility(1,1): Error: common|chrome/editor.yaml:214 must define BookmarkRestoreKeyPrefix if BookmarkKeyCount > 0.
```

rob-v commented:
Could you clarify the logic behind hotkeys defined in Widget vs Logic due to different configuration/implementation?

pchote commented:
Keys that control or interact with a widget go on that widget.  Everything else should be implemented as stand-alone logic classes.

reaperrr commented:
Note: it doesn't seem to be caused by this PR (hence the +1), but next/prev tab (page u/down) don't work on my system, neither with this PR nor on bleed.

Everything else works, but I'll wait for a reply (and rebase, obviously) before I merge this.

rob-v commented:
Page up/down works e.g. in TD to switch between all barracks,... iirc.

reaperrr commented:
> Page up/down works e.g. in TD to switch between all barracks,... iirc.

Ah, I see. That works, so just needs a rebase then.

pchote commented:
Rebased.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3b138dbf9bd34d319974cf8c92790a73878d6d8f...00ab906438192ab9831ea68647580601a1283314
